implementations/rust/ockam/ockam_entity/src/profile_set.rs

```diff
@@ -11,62 +11,125 @@ pub use identifiers::*;
 pub mod secrets;
 pub use secrets::*;
 
+use crate::EntityError::{InvalidInternalState, InvalidParameter};
+use ockam_core::hashbrown::hash_map::HashMap;
+
 /// An Entity represents an identity in various authentication contexts.
 #[derive(Clone)]
-pub struct Entity<P: ProfileTrait> {
+pub struct ProfileSet<P: ProfileTrait> {
     default_profile_identifier: ProfileIdentifier,
-    profiles: Vec<P>,
+    profiles: HashMap<ProfileIdentifier, P>,
+}
+
+pub trait ProfileRetrieve<P: ProfileTrait> {
+    fn profile(&self, profile_identifier: &ProfileIdentifier) -> Option<&P>;
+}
+
+pub trait ProfileAdd<P: ProfileTrait> {
+    fn add_profile(&mut self, profile: P) -> Result<()>;
+}
+
+pub trait ProfileUpdate<P: ProfileTrait> {
+    fn update_profile(&mut self, old_profile_id: &ProfileIdentifier, profile: P) -> Result<()>;
```

Comment:
1. Profile identifier doesn't change (that's why I think there is no such thing as "old_profile_id")
2. I think we only update profile with new set of change events

Reply:
I was wondering if update was necessary at all - removing for now